[AOTInductor] [Tooling] Update NaN and INF Checker for AOTInductor

muchulee8 opened the pull request:
Summary:
1. Integrate NaN and INF checker with existing config, controllable by env var.
2. Move inject point of NaN & INF checker earlier, this could prevent buffer freeing before check.
3. Inject debugging code in Kernel level, which prevents us trying to read buffers that are fused inplace and into a single kernel.

Test Plan:
Debugging utility.
Test and check by existing tests with env var:
```
TORCHINDUCTOR_NAN_ASSERTS=1 TORCHINDUCTOR_MAX_AUTOTUNE=0 python test/inductor/test_aot_inductor.py -k AOTInductorTestNonABICompatibleCuda.test_seq_non_abi_compatible_cuda
```

Reviewed By: ColinPeppler

Differential Revision: D57989176


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @ColinPeppler @amjames @desertfire @chauhang

chenyang78 commented on torch/csrc/inductor/aoti_torch/utils.h:
```diff
@@ -48,6 +48,17 @@ inline AtenTensorHandle new_tensor_handle(at::Tensor&& tensor) {
+  for (int64_t i = 0; i < flattened.numel(); i++) {
+    auto value = flattened[i].item<float>();
+    if (std::isinf(value) || std::isnan(value)) {
+      assert(false);
```
nit - can we throw an exception? ```assert``` may become a no-op if we compile the code in non-debug mode.

muchulee8 replied:
changed to throw.

chenyang78 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -770,15 +770,21 @@ AOTITorchError aoti_torch_repeat_interleave_Tensor(
 
 // Function to check existence of inf and NaN
-AOTITorchError aoti_check_inf_and_nan(AtenTensorHandle tensor) {
+AOTITorchError aoti_check_inf_and_nan(
```
nit - this function looks very similar to another one (i.e. ```assert_inf_and_nan```) defined in utils.h. Could we only keep one?

desertfire replied:
ditto on this one. `assert_inf_and_nan` seems only exists because of the non-ABI compatible mode. We can either ignore that mode or just call `assert_inf_and_nan` here.

muchulee8 replied:
Resolved. Let the ABI-compat mode directly call into utils.

muchulee8 commented:
@pytorchbot label "topic: not user facing"

muchulee8 commented:
@pytorchbot merge

muchulee8 committed:
[AOTInductor] [Tooling] Update NaN and INF Checker for AOTInductor (#127574)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/127574

1. Integrate NaN and INF checker with existing config, controllable by env var.
2. Move inject point of NaN & INF checker earlier, this could prevent buffer freeing before check.
3. Inject debugging code in Kernel level, which prevents us trying to read buffers that are fused inplace and into a single kernel.

Test Plan:
Debugging utility.
Test and check by existing tests with env var:
```
TORCHINDUCTOR_NAN_ASSERTS=1 TORCHINDUCTOR_MAX_AUTOTUNE=0 python test/inductor/test_aot_inductor.py -k AOTInductorTestNonABICompatibleCuda.test_seq_non_abi_compatible_cuda```

Reviewed By: ColinPeppler

Differential Revision: D57989176

muchulee8 commented:
@pytorchbot merge

muchulee8 commented:
@pytorchbot merge
